core/store/src/trie/trie_storage.rs

```diff
@@ -184,30 +245,90 @@ impl TrieCachingStorage {
     fn inc_counter(&self) {
         self.counter.set(self.counter.get() + 1);
     }
+
+    /// Get position of the given key in the cache.
+    pub(crate) fn get_cache_position(&self, key: &CryptoHash) -> CachePosition {
+        match self.chunk_cache.borrow_mut().get(key) {
+            Some(value) => CachePosition::ChunkCache(value.clone()),
+            None => match self.shard_cache.get(key) {
+                Some(value) => CachePosition::ShardCache(value.clone()),
+                None => CachePosition::None,
+            },
+        }
+    }
+
+    /// If we are in the caching chunk state, put item to the chunk cache.
+    fn try_put_to_chunk_cache(&self, key: CryptoHash, value: Arc<[u8]>) {
+        if let TrieCacheState::CachingChunk = self.cache_state.borrow().get() {
+            self.chunk_cache.borrow_mut().insert(key, value);
+        };
+    }
+
+    /// Get value for the given key from cache.
+    /// Return value if it is present and cost of retrieving the trie node corresponding to this key.
+    pub(crate) fn get_from_cache(&self, key: &CryptoHash) -> RawBytesWithCost {
+        match self.get_cache_position(key) {
+            // If no value is present, return None. If node for this key will be retrieved from DB, we have to charge
+            // full cost for it.
+            CachePosition::None => {
+                RawBytesWithCost { value: None, cost: TrieNodeRetrievalCost::Full }
+            }
+            // If value is present only in shard cache, we can copy value to the chunk cache. The cost must be full
+            // anyway because we didn't access the node in the chunk yet.
+            CachePosition::ShardCache(value) => {
+                self.try_put_to_chunk_cache(key.clone(), value.clone());
+                RawBytesWithCost { value: Some(value), cost: TrieNodeRetrievalCost::Full }
+            }
+            // If value is present in chunk cache, the cost is free.
+            CachePosition::ChunkCache(value) => {
+                RawBytesWithCost { value: Some(value), cost: TrieNodeRetrievalCost::Free }
+            }
+        }
+    }
+
+    /// Put a key-value pair into cache.
+    pub(crate) fn put_to_cache(&self, key: CryptoHash, value: Arc<[u8]>) {
+        self.shard_cache.put(key.clone(), value.clone());
+        self.try_put_to_chunk_cache(key, value.clone());
+    }
+
+    /// Set cache state.
+    pub fn set_state(&self, state: TrieCacheState) {
```

Comment:
Added a comment with call into `execute_function_call`. I uncomment it in the following PR with the protocol upgrade.

The cache is dropped when `TrieCachingStorage` will be dropped. It is created in the beginning of the `apply_transactions_with_optional_storage_proof`, and dropped in the end.

Reply:
why do we need to switch between the two mode/states? How will `TrieCachingStorage` be used outside of `apply_transactions_with_optional_storage_proof`?

Reply:
Instead of use `set_state`, how about we use `start_chunk_cache` and `finish_chunk_cache` to indicate the state(mode) transition. I think it's clearer that way for what the switch is for and where these methods should be called. And you also need to clear the chunk cache in `finish_chunk_cache`

Reply:
> why do we need to switch between the two mode/states?

Ideally we would work only in `CachingChunk` mode. Unfortunately, there are many trie node touches for which we don't charge any costs (caused by `get_account`, `get_access_key`, etc.). So we enable this mode only during funcall processing, because all trie touches are charged there. It allows us to imply that cache chunk size is limited by ~31K.

> How will `TrieCachingStorage` be used outside of `apply_transactions_with_optional_storage_proof`?

Not sure I understood the question. `TrieCachingStorage` is created in the `get_trie_for_shard`, and is removed when the call ends. So its lifetime is similar to `apply_transactions_with_optional_storage_proof` call.

Reply:
> Instead of use set_state, how about we use start_chunk_cache and finish_chunk_cache to indicate the state(mode) transition. I think it's clearer that way for what the switch is for and where these methods should be called. 

I agree, but with the current architecture we need to create "copies" of these methods in `RuntimeExt` and `TrieUpdate` to access the state from the outside. I think the current approach minimizes the introduced ugliness.

> And you also need to clear the chunk cache in finish_chunk_cache

I don't. This cache stores nodes for the whole chunk, so following receipts could spend less gas. It is done intentionally, and AFAIK it was discussed on some protocol meeting.